helix-rest/src/main/java/org/apache/helix/rest/metadatastore/accessor/ZkRoutingDataWriter.java

```diff
@@ -87,16 +87,18 @@ public ZkRoutingDataWriter(String namespace, String zkAddress) {
 
     // Get the hostname (REST endpoint) from System property
     String hostName = System.getProperty(MetadataStoreRoutingConstants.MSDS_SERVER_HOSTNAME_KEY);
+    String port = System.getProperty(MetadataStoreRoutingConstants.MSDS_SERVER_PORT_KEY);
     if (hostName == null || hostName.isEmpty()) {
       throw new IllegalStateException(
-          "Unable to get the hostname of this server instance. System.getProperty fails to fetch "
+          "Unable to get the hostname of this server instance or the hostname is empty. System.getProperty fails to fetch "
               + MetadataStoreRoutingConstants.MSDS_SERVER_HOSTNAME_KEY + ".");
     }
-    // remove trailing slash
-    if (hostName.charAt(hostName.length() - 1) == '/') {
-      hostName = hostName.substring(0, hostName.length() - 1);
+    if (port == null || port.isEmpty()) {
+      throw new IllegalStateException(
+          "Unable to get the port of this server instance or the port is empty. System.getProperty fails to fetch "
+              + MetadataStoreRoutingConstants.MSDS_SERVER_PORT_KEY + ".");
     }
-    _myHostName = HttpConstants.HTTP_PROTOCOL_PREFIX + hostName;
+    _myHostName = HttpConstants.HTTP_PROTOCOL_PREFIX + hostName + ":" + port;
     ZNRecord myServerInfo = new ZNRecord(_myHostName);
```

Comment:
Using a vanilla hostname would be enough here. Let's create a ZNRecord with two simple fields added: hostname and port?

Reply:
updated with a znode with 3 fields, as we discussed.